### DISCUSSION TITLE
Issue with conform.nvim autoformat in Neovim (Insert Mode + <C-s> Shortcut)

### DISCUSSION THREAD

**Discussion by MatejBransky** [LazyVim/LazyVim#4600] [Q&A]
Hi everyone,

I’m encountering a strange issue with conform.nvim and ~~Prettier (enabled in LazyExtras)~~ _(it's probably not related to the Prettier, see [the finding below](https://github.com/LazyVim/LazyVim/discussions/4600#discussioncomment-10973483))_ when using the save shortcut `<C-s>` for saving the buffer in Neovim.

The problem happens specifically in insert mode when I press `<C-s>` on an empty line that is above a line with content. Here’s what happens:

The empty line is correctly deleted (which is expected).
The buffer is saved.
However, the line with content below the deleted line gets modified—its indentation is removed, leaving the buffer in a modified state.
After this happens, if I hit `<C-s>` again, the formatting is corrected, and the buffer is saved without further changes.

I’ve linked a [recording of this behavior](https://youtu.be/I0byoeD7uco). Also, I’m using the default initial setup of the LazyVim distribution, without any changes to the configuration.

Does anyone know what could be causing this behavior?

Thanks for any help!

**Comment by dotfrag** [comment]
I have had similar problems with autoformatting, where saving once would reset indentation on one line, and then saving twice fixed it.. Can't remember the language/formatter combination though.

May I ask why you use prettier for lua?

**Reply by MatejBransky** [reply]
I added Prettier for TypeScript at work. But I noticed it handles other programming languages as well, so I don’t stop it from doing so. However, I’m having the same issue in other languages, including TypeScript.

I’ve just found out that if I call :write or :update directly without `<Esc>`, the problem doesn’t occur. So it seems that the following `<Esc>` command (see LazyVim keymap for the `<C-s>` [here](https://github.com/LazyVim/LazyVim/blob/13a4a84e3485a36e64055365665a45dc82b6bf71/lua/lazyvim/config/keymaps.lua#L69)) somehow disrupts the formatted and already saved buffer.

I’ll try, for the time being, to work with just saving without switching to normal mode using `<Esc>`.

**Reply by dotfrag** [reply]
I don't believe prettier is used for lua, to verify that, open a lua file and run `:ConformInfo`.

**Reply by MatejBransky** [reply]
I'm starting to think it might be related to something else on my MacBook. Because before yesterday's reinstallation of OS X, I didn't have this issue. 

However, I would have expected that after a fresh install, with no unnecessary stuff around, it would actually work better than before. 😅

**Reply by dotfrag** [reply]
I think the same, but I can't think what might cause it.

**Reply by MatejBransky** [reply]
> I don't believe prettier is used for lua, to verify that, open a lua file and run :ConformInfo.

You,re right. So it's not probably related to the Prettier. Good to know. 👍 

```
Log file: /Users/matej.bransky/.local/state/nvim/conform.log
          additional information: expected '}'

Formatters for this buffer:
LSP: lua_ls
stylua ready (lua) /Users/matej.bransky/.local/share/nvim/mason/bin/stylua

Other formatters:
fish_indent unavailable: Command not found
markdown-toc unavailable: Condition failed
markdownlint-cli2 unavailable: Condition failed
prettier unavailable: Condition failed
shfmt ready (sh) /Users/matej.bransky/.local/share/nvim/mason/bin/shfmt
```

**Reply by dotfrag** [reply]
Yes, it will use `stylua` formatting.

**Reply by MatejBransky** [reply]
I mistakenly linked it to Prettier because I encountered the issue while working in TypeScript, and when I removed it from LazyExtras, the problem went away, but of course, it didn't format either.

So the issue is with the subsequent `<Esc>`/`<C-c>` when switching from insert mode to normal mode after the autoformatting. See the screen recording. _(The symbol `ß` should be `<M-s>` which I've configured for the `<Cmd>:update<CR>` command._

https://github.com/user-attachments/assets/6315bf01-a4a2-4499-8dad-cab3f3052841

**Reply by MatejBransky** [reply]
So it's not related to `<Esc>`/`<C-c>` either. It also happens when I simply move the cursor with the arrow keys while in insert mode.

Conditions for reproduction:

- I have to be on a line that will likely be deleted as part of the formatting,
- below this line, there must be an indented line with content — this line should remain indented after formatting,
- I have to move the cursor up or down.

https://youtu.be/74lDIoQhv7A

**Reply by dpetka2001** [reply]
I'm not sure why this happens, but here is my guess. 

When you save in insert mode the cursor goes at the very first column of the line, not the first character of the line. When you hit `<esc>` the line gets deindented to where the cursor is. When you save again, of course the formatter formats the text as it should in normal mode. 

From my experimenting there are 2 workarounds to this. 

The first and most obvious one is to call `<esc>` at the start of the command instead of the end. That produces a little visual glitch that you first go into normal mode and then the formatting happens. 

The second is to remap the key to `map({ "i", "x", "n", "s" }, "<C-s>", "<cmd>execute 'write' |normal ^<cr><esc>", { desc = "Save File" })
`. This correctly formats even in insert mode because it moves the cursor to the first character of the line before exiting insert mode. A little side effect of this is that even if you are in normal mode it will move the cursor from its current position to the first character of the line which might not be what a user wants. For the second workaround it might be preferable to only apply this mapping in insert mode and keep the original for the rest of the modes.

Not sure what the best and more desirable solution is.

PS: I personally like the second workaround better, because there's no visual glitch that you go to insert mode and then the formatting happens. That's probably what I will add to my personal mappings.

PS2: Of course when you're insert mode and not on an empty line, then your cursor will still move to the first character of the line, which might be somewhat undesirable. I believe the `<esc>` first might be the most desirable solution, even if it provides a little visual glitch.

**Reply by dotfrag** [reply]
Thanks @dpetka2001, I now have this in my config and formatting works as expected.

```lua
map({ "i", "x", "n", "s" }, "<C-s>", "<esc><cmd>w<cr>", { desc = "Save File" })
```
